### PR TITLE
Add closed pinned issues notification on Slack

### PR DESCRIPTION
**Pull request by mollyIV**
### Motivation and Context

Recently the most important issues are pinned to the main page in order to easily follow the progress and keep focus on them. However it's quite easy to get an outdated state by closing the issues and not unpinning them. This PR is introducing Slack notifications whenever someone closes the pinned issue - in that way a contributor will follow a link and unpin the issue right away 🎉 

### Implementation Details

Getting [pinned issues](https://developer.github.com/v4/object/pinnedissue/) is available since [GitHub GraphQL API v4](https://developer.github.com/v4/) (although still in preview). 

We added a new client [graphql-client](https://github.com/github/graphql-client) dependency (developed by GitHub) to communicate with graphQL API easily. 

I added the `schema.json` JSON file to the repo as well as suggested by the documentation:

```
  # However, it's smart to dump this to a JSON file and load from disk
  #
  # Run it from a script or rake task
  #   GraphQL::Client.dump_schema(SWAPI::HTTP, "path/to/schema.json")
  #
  # Schema = GraphQL::Client.load_schema("path/to/schema.json")
```

I have been testing those changes on my private repo, here are the results:

screenshot 2019-02-28 at 06 47 08

screenshot 2019-02-28 at 06 47 58

### Future improvements

💡  we could unpin the issue automatically and just send an info notification on Slack.
💡 recently we have one entry point in the code `def start(process: :issues)` which points that we do not leverage the power of [webhooks](https://developer.github.com/webhooks/). It would be great to execute the action by bot right away (i.e. new PR created, issues has been closed, ...) instead of waiting given interval of time.
💡 at some point we could migrate to [App](https://developer.github.com/apps/) which is recommended way of GitHub integration.

### CODE REVIEW

**Comment by janpio** [comment]
I don't understand any of the code, but I love what it does :)

File: Rakefile [hunk 1]
```diff
@@ -68,6 +68,20 @@ task :post_unreleased_changes do
...
 end
 
+desc 'Dump the schema for a v4 Github GraphQL API'
+task :dump_schema do
+  logging_exceptions('dump-schema.log') do |logger|
+    GitHubAPI.dump_schema()
+  end
+end
+
+desc 'Post incoming regression issues on Slack'
```
Comment (mollyIV) [inline Rakefile]: The description of the task is wrong 😬

File: bot.rb [hunk 2]
```diff
@@ -334,6 +340,37 @@ def process_code_signing(issue)
...
+        if pinned_issue.closed
+          slack_about_closed_pinned_issues(pinned_issue)
+        end
+      end
+    end
+
+    def slack_about_closed_pinned_issues(pinned_issue)
+      return unless Bot.should_send_trivial_slack_notification?
+
+      logger.info("Notifying the Slack room about pinned issue that is closed...")
```
Comment (dantoml) [inline bot.rb]: could we include the issue number in the log?

File: bot.rb [hunk 3]
```diff
@@ -334,6 +340,37 @@ def process_code_signing(issue)
...
+
+      post_body = {
+        text: "Caution: Item ##{pinned_issue.number} is still a pinned issue although it was recently closed: #{pinned_issue.url}"
+      }.to_json
+
+      logger.info(post_body)
```
Comment (dantoml) [inline bot.rb]: ❓ I assume this was for debugging, could we remove it now? (to avoid logging twice)

File: bot.rb [hunk 4]
```diff
@@ -334,6 +340,37 @@ def process_code_signing(issue)
...
+
+      response = Excon.post(ACTION_CHANNEL_SLACK_WEB_HOOK_URL, body: post_body, headers: { "Content-Type" => "application/json" })
+
+      if response.status == 200
+       logger.info("Successfully notified the Slack room about pinned issue that is closed")
+      else
+       logger.info("Failed to notify the Slack room about pinned issue that is closed")
```
Comment (dantoml) [inline bot.rb]: Maybe log the `response.status` and `response.body` here to allow easier debugging?

File: bot.rb (hunk 3 above)
Comment (mollyIV) [reply bot.rb]: Absolutely, I am gonna remove it, thanks 👍

**Comment by mollyIV** [comment]
Hey @dantoml 👋 Thanks a bunch for a code review 🙏 

> I'm curious as to why we're doing graphql directly rather than using oktokit though? (guessing because filtering by pinned doesn't look obvious in the gem docs)

`octokit` does support GitHub API **v3** however it does not support [graphql API (**v4**)](https://developer.github.com/v4/). There are [graphql.js](https://github.com/octokit/graphql.js) and [graphql.net](https://github.com/octokit/octokit.graphql.net) libraries but there's no ruby one. 

That is the reason ☝️ why we decided doing graphql calls directly 🙂

**Comment by mollyIV** [comment]
💡When I checked if there's `octokit graphql.ruby` library, I found [`graphql-schema` repo](https://github.com/octokit/graphql-schema) so maybe instead of adding `schema.json` by our own (it's a huge file), we could use and restrict the version:

```ruby
const {schema} = require('@octokit/graphql-schema')
schema.json // JSON version
```

WDYT @dantoml ? 🤔

**Comment by mollyIV** [comment]
@dantoml I have made the requested changes, could you please have a 👀🙇🏻‍♂️

**Pull request merged**